rfcs/20220329-Extending-CUDA-backend-of-oneDNN-Primitives/README.md

```diff
@@ -0,0 +1,190 @@
+
+ - This contribution would extend oneDNN's support for more primitives in CUDA backend. 
+
+ - Wherever direct mapping of primitives from oneDNN is not available in cuDNN, the implementation would proceed with custom kernels for emulating their functionality on the backend device. 
+
+ 
+
+## Supported Primitives,  Implementation Limitations and Proposed Solution 
+
+ 
+
+### Reduction 
+
+- Reduction primitive in CUDA backend can be implemented using cudnnReduceTensor primitive of cuDNN.
+- cudnnReduceTensor supports only five modes of reduction operations via enumerators which are as follows.  
+    - `CUDNN_REDUCE_TENSOR_MAX` 
+    - `CUDNN_REDUCE_TENSOR_MIN` 
+    - `CUDNN_REDUCE_TENSOR_MUL`
+    - `CUDNN_REDUCE_TENSOR_ADD` 
+    - `CUDNN_REDUCE_TENSOR_AVG` 
+
+- Supported Data Types:  f32, s8 
+
+ 
+
+#### Implementation Limitations 
+
+- Lack of support for the algorithms of  
+
+    - `reduction_norm_lp_max` 
+
+    - `reduction_norm_lp_sum` 
+
+    - `reduction_norm_lp_power_p_max` 
+
+    - `reduction_norm_lp_power_p_sum` 
+
+- No support for bf16 datatype. 
+
+ 
+
+#### Proposed Solutions 
+
+- To achieve the functionality of `reduction_norm_lp_power_p_max` and `reduction_norm_lp_power_p_sum` algorithms in cuDNN, `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` can be used in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 1.  
+
+- Similarly, the functionality of `reduction_norm_lp_max` and `reduction_norm_lp_sum` algortihms can be implemented using `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 2.  
+
+- As an alternative to the above workarounds, the functionality of the above algorithms can be implemented through a custom kernel in SYCL. 
+
+ 
+
+### Recurrent Neural Networks 
+
+- RNN primitive in CUDA backend can be implemented using `cudnnRNNForward` and `cudnnRNNBackward` primitives of cuDNN. 
+
+- cuDNN's primitives only support two algorithm of RNN. 
+
+    - `CUDNN_LSTM` 
+
+    - `CUDNN_GRU` 
+
+- Supported Data Types: f16, f32 
+
+ 
+
+#### Propagation 
+
+Supported propagation kinds: Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+- No support for following algorithms in cuDNN. 
+
+    - `augru` 
+
+    - `lbr_augru` 
+
+    - `lbr_gru` 
+
+    - `vanilla_rnn` 
+
+- No support for U8, bf16 datatypes. 
+
+ 
+
+#### Proposed Solution 
+
+The unsupported algorithms of `augru`, `lbr_augru`, `lbr_gru`, `vanilla_rnn` in cuDNN can be implemented using custom kernels, per the mathematical expressions of the algorithms. 
```

Comment:
As I mentioned in a synch-up meeting GEMM is the main performance ingredient in RNN so cuBLAS can greatly help with this.